src/Avalonia.Base/Media/Fonts/Tables/Cmap/CmapFormat12Table.cs

```diff
@@ -0,0 +1,183 @@
+﻿using System;
+using System.Buffers.Binary;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace Avalonia.Media.Fonts.Tables.Cmap
+{
+    internal sealed class CmapFormat12Table : IReadOnlyDictionary<int, ushort>
+    {
+        private readonly ReadOnlyMemory<byte> _table;
+        private readonly int _groupCount;
+        private readonly ReadOnlyMemory<byte> _groups;
+
+        private int? _count;
+
+        public CmapFormat12Table(ReadOnlyMemory<byte> table)
+        {
+            var reader = new BigEndianBinaryReader(table.Span);
+
+            ushort format = reader.ReadUInt16();
+            Debug.Assert(format == 12, "Format must be 12.");
+
+            ushort reserved = reader.ReadUInt16();
+            Debug.Assert(reserved == 0, "Reserved field must be 0.");
+
+            uint length = reader.ReadUInt32();
+
+            _table = table.Slice(0, (int)length);
+
+            uint language = reader.ReadUInt32();
+
+            _groupCount = (int)reader.ReadUInt32();
+
+            int groupsOffset = reader.Position;
+            int groupsLength = _groupCount * 12;
+
+            Debug.Assert(length >= groupsOffset + groupsLength, "Length must cover all groups.");
+
+            _groups = _table.Slice(groupsOffset, groupsLength);
+        }
+
+        private static uint ReadUInt32BE(ReadOnlyMemory<byte> mem, int groupIndex, int fieldOffset)
```

Comment:
Pass a `ReadOnlySpan` here for performance and make the callers pass the span instead (see comment below).